Teachable learnset helper mechanics

Bassoonian opened the pull request:
This tool, which runs when building a ROM, does the following:
- Scans the repo for TMs and tutor moves
- Checks json files (from PoryMoves, with minor label adjustments for an easier time) for compatibility
- Creates a `teachable_learnsets.h` file with the smallest size possible by only including the moves found in the scan

Other features:
- Users can easily delete/add jsons for less/more data
- Any changes in `teachable_learnsets.h` that were made before this tool runs for the first time will be saved in `custom.json` for flawless migration

Fixes #1379 
Fixes #2274

## Internal to do list
- [x] list moves found in scan
- [x] integrate in Makefile
- [x] clarify python instructions in Install.md
- [x] update sv.json based on The Indigo Disk data ([source](https://gist.githubusercontent.com/SciresM/a1a0c2e48ae5ca58a0ecd131bb7f7845/raw/f7d5d4fc4c0de9420030421aa8d8b64fc5dbca22/personal.txt))

## Necessities before merging
- [x] #4026 needs to be merged into upcoming
- [x] #4027 needs to be resolved and (after master) be merged into upcoming

## What this PR does not contain
- XD data is currently limited to Deoxys' Normal form (as it contains compatibilities missing from all subsequent games)

## **Discord contact info**
bassoonian

mrgriffin commented on tools/teachablelearnsethelper/teachable_learnset_helper.py:
```diff
@@ -0,0 +1,89 @@
+for file in incs_to_check:
+    with open(file, 'r') as f2:
+        raw = f2.read()
+    if 'special ChooseMonForMoveTutor' in raw:
+        for x in re.findall('setvar VAR_0x8005, (MOVE_.*)', raw):
+            if not x in tutor_moves:
+                tutor_moves.append(x)
```
I think there's a few ways in which is this a little weird:
1. Somebody might use `setvar VAR_0x8005, MOVE_...` for something other than `ChooseMonForMoveTutor`, e.g. they might have a special that teaches a particular move (which is in `VAR_0x8005`, a sensible choice because `VAR_0x8004` contains the result of `ChoosePartyMon`)..
2. If somebody did, e.g. `copyvar VAR_0x8005, VAR_MOVE_TO_TEACH` then it won't be detected (and they'd probably be doing that in a way that makes it difficult to know what `VAR_MOVE_TO_TEACH` contains, at the very least through a `call`, but it could be a variable set in an entirely different script). Related, (and somewhat more likely) somebody could write a special which writes the move ID into `gSpecialVars_0x8005`.
3. If you write `@ setvar VAR_0x8005, MOVE_TACKLE` I think that would be picked up. This is both good and bad, bad because commenting things out should work, but good because it gives a way to explicitly list teachable moves in cases where it can't be picked up automatically (like 2).
4. Very, very minor, but this will find scripts that aren't OW scripts in `data/scripts` (i.e. not referenced by `data/event_scripts.s`). This is very unlikely to matter unless people write other script languages that can also reference OW vars.
5. If somebody happened to write move tutors in a macro they won't be detected. (I personally think a macro is a bad idea because I want to see what's happening to the special vars, but I could see other people thinking differently and could easily see a change like that making it into here or pret\*).

I don't have a workable suggestion for these because I think you'd have to do static analysis to work out which value is in `VAR_0x8005` when `special ChooseMonForMoveTutor` runs. In practice it probably won't come up except for people who are really pushing the boundaries of the scripting language. afaict MoT wouldn't see any bugs, although we do hit 1-3 points above and almost 4 (we just happen not to ever assign a move to `VAR_0x8005`).

tl;dr not using a parser, but I don't expect it to cause a problem. It's worth noting that something similar to 2 exists in vanilla: [`GetBattleFrontierTutorMoveIndex`](<https://github.com/pret/pokeemerald/blob/6c96641035099db8a2c1e680fb8793f1b29cf2fd/src/field_specials.c#L3188>) (it generates an index rather than a move ID because vanilla's tutors take indices not move IDs).

btw, I'd like to be really explicit that I love the idea of determining the smallest possible data structures at build time.

\* Not pret for specifically move tutors, because they work differently there.

GraionDilach commented:
> Creates a teachable_learnsets.h file with the smallest size possible by only including the moves found in the scan

Not necessarily, because the generated header lacks the species_enabled gates. This means that with enough disabled species, the currently shipped alternative will be smaller than the generated result.

Bassoonian commented:
> > Creates a teachable_learnsets.h file with the smallest size possible by only including the moves found in the scan
> 
> Not necessarily, because the generated header lacks the species_enabled gates. This means that with enough disabled species, the currently shipped alternative will be smaller than the generated result.

I'm unsure what you mean. The file generated preserves the P_FAMILY checks, so even if the generated result would be larger than the currently shipped .h file (which is technically _too_ small because it lacks some learnsets and compatibilities), all the disabled species wouldn't be compiled in the game. The final result is still the smallest possible list for each species (with only the enabled species being compiled, just as the vanilla file).

GraionDilach commented:
Hm, my python knowledge failed here. I was quite certain this strips away the preprocessor lines, but my test show it doesn't. I don't understand that part, but okay.

Bassoonian commented:
It simply regexes the existing `static const u16 s<speciesnamehere>TeachableLearnset[]`` without touching what actually surrounds them, so the family defines (and comments, if the user would for some reason want to add comments to the file, and #defines, etc) are preserved in their original position

Bassoonian commented:
This PR is now ready for review. Please make sure to check the Python instructions in the installation instructions properly, as I have very little (read: no) experience outside of OSX, and even on my own computer I forgot how I installed it :)

AlexOn1ine commented:
Is there a point in adding a `custom.json` file that has only the species data without the moves for users so they know they can add new moves to it?

Bassoonian commented:
I don't think so, considering the thing will make a `custom.json` for you if there's any non-vanilla data in the file when you first run it. You can also name it whatever as long as it's a json file, so there could be a `gaia.json` or an `unbound.json` or whatever you'd prefer.

AsparagusEduardo commented:
I tested by removing this:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/7f02cae2-96a5-4735-a7ac-39db9d7d9142)

And now it's not building, with no information about the error:
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/5c326750-75c1-4ca9-bda1-aedddfdc93f5)

Also, I think it'd be better to have the "Updated" message formatted like this:
```
Updated Teachable Learnsets for: Charmander, Charmeleon, Charizard, Squirtle, ...
```

Bassoonian commented:
Oh, huh, interesting. It's refusing to print anything when triggered by Makefile on my end, while it seems to ignore newlines in yours. I think that the error is quite literally what it says, i.e. the print is being considered a command by bash? Do you get the same error if you just run it without removing the Dynamic Punch line?

AsparagusEduardo commented:
> Do you get the same error if you just run it without removing the Dynamic Punch line?
I first ran make with the branch as-is. I removed the line afterwards and did make.

Restoring the line doesn't fix it.
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/47ecff47-0ac5-47e6-a55d-40c3394cd02d)

AsparagusEduardo commented:
> Oh, huh, interesting. It's refusing to print anything when triggered by Makefile on my end, while it seems to ignore newlines in yours. I think that the error is quite literally what it says, i.e. the print is being considered a command by bash?

I think it'll be better to just separate this as its own tool instead of having it part of makefile, which was something I requested previously.

Bassoonian commented:
We previously reached a majority vote on just integrating it, since if it's not built into the makefile it may just as well not be a part of the repo. It worked previously, I don't know what broke it, so I'll try to fix it

AsparagusEduardo commented:
> We previously reached a majority vote on just integrating it, since if it's not built into the makefile it may just as well not be a part of the repo. It worked previously, I don't know what broke it, so I'll try to fix it

Where was the vote?

Bassoonian commented on Makefile:
```diff
@@ -461,6 +462,10 @@ $(OBJ_DIR)/sym_common.ld: sym_common.txt $(C_OBJS) $(wildcard common_syms/*.txt)
 
+# NOTE: Depending on event_scripts.o is hacky, but we want to depend on everything event_scripts.s depends on without having to alter scaninc
+$(DATA_SRC_SUBDIR)/pokemon/teachable_learnsets.h: $(DATA_ASM_BUILDDIR)/event_scripts.o
+	$(TEACHABLELEARNSETHELPER)
```
This solved the issue indeed, both the (occasional lack of) printing and the printing being interpreted by bash. Thanks! :)

Bassoonian commented:
Fixed Makefile (thanks MGriffin!), expanded the header to refer to the Python file and added a config to easily disable the helper without having to fiddle with Makefile. Ready for re-review!

AsparagusEduardo commented on include/config/pokemon.h:
```diff
@@ -37,6 +37,9 @@
 #define P_TM_LITERACY               GEN_LATEST  // Since Gen 6, TM illiterate Pokémon can learn TMs that teach moves that are in their level-up learnsets.
 
+// Learnset helper toggles
+#define P_LEARNSET_HELPER_TEACHABLE TRUE        // If TRUE, teachable_learnsets.h will be populated by tools/learnset_helpers/teachable.py using the included json files based on available TMs and tutors.
```
```suggestion
#define P_LEARNSET_HELPER_TEACHABLE TRUE        // If TRUE, teachable_learnsets.h will be populated by tools/learnset_helpers/teachable.py using the included JSON files based on available TMs and tutors.
```

AsparagusEduardo commented on tools/learnset_helpers/teachable.py:
```diff
@@ -3,6 +3,11 @@
+# before all else, abort if the config is off
+with open("./include/config/pokemon.h", "r") as file:
+    if re.findall("#define P_LEARNSET_HELPER_TEACHABLE *([^ ]*)", file.read())[0] != "TRUE":
+        quit()
```
There's no handling in case `P_LEARNSET_HELPER_TEACHABLE` doesn't exist.
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/22fa59c3-d0bb-48f0-b4f7-ab41b752a520)

Bassoonian replied:
Good catch!

AsparagusEduardo reviewed (changes requested):
We're almost there
